### PR TITLE
Fix #570: add support for Brush customizations.

### LINKED ISSUE #570: Support `Brush` customizations
If we add support for `Brush` and `() -> Brush` customizations, then on Android > T we can replace parts of the UI with fragment shaders.

Initially, adding support for `Brush` allows us to test out the infrastructure. Later on, we can figure out how to let the designer specify the fragment shader and its uniform values.

### PR DESCRIPTION
**Pull request by iamralpht**
This change adds basic support for customizing the `Brush` that a frame renders with.

### CODE REVIEW

**Comment by iamralpht** [comment]
Let's see what CI does with this! I haven't added Brush customizations for text yet, but I'll add that later this evening.

**Comment by iamralpht** [comment]
@rylin8 I think this change is ready for your review. I am wondering:

- should the `BrushFunction` take an `ImageCustomizationContext`? Then someone could put the color in as a shader uniform and make some common shaders that they might use for different backgrounds (from different docs).
- text doesn't seem to update live; I'm not sure what's up with that... it's good enough for a first pass but we'll need a ticket to come back.

File: docs/_docs/modifiers/modifiers.md [hunk 1]
```diff
@@ -122,6 +122,57 @@ called `artwork`, which is a nullable `android.graphics.Bitmap` instance. To use
 the background fill from the document, pass a value to replace the background
 fill or pass `null`.
 
+### Brush content {#brush-content}
```
Comment (rylin8) [inline docs/_docs/modifiers/modifiers.md:125]: Wow documentation too!

File: designcompose/src/main/java/com/android/designcompose/FrameRender.kt [hunk 2]
```diff
@@ -318,18 +318,33 @@ internal fun ContentDrawScope.render(
...
             val p = Paint()
-            val b = background.asBrush(document, density)
-            if (b != null) {
-                val (brush, fillOpacity) = b
-                brush.applyTo(frameSize, p, fillOpacity)
-                p
-            } else {
-                null
+            customFillBrush.applyTo(frameSize, p, 1.0f)
+            listOf(p)
```
Comment (rylin8) [inline designcompose/src/main/java/com/android/designcompose/FrameRender.kt:333]: That's pretty simple.

File: integration-tests/validation/src/main/java/com/android/designcompose/testapp/validation/MainActivity.kt [hunk 3]
```diff
@@ -1761,6 +1777,78 @@ fun VariableBorderTest() {
     VariableBorderTestDoc.Main()
 }
 
+@DesignDoc(id = "oetCBVw8gCAxmCNllXx7zO")
```
Comment (rylin8) [inline integration-tests/validation/src/main/java/com/android/designcompose/testapp/validation/MainActivity.kt:1780]: @timothyfroehlich I thought these were in separate files now, or is that only in the layout branch? If so what was the reason for that?

**Comment by rylin8** [comment]
> @rylin8 I think this change is ready for your review. I am wondering:
> 
> * should the `BrushFunction` take an `ImageCustomizationContext`? Then someone could put the color in as a shader uniform and make some common shaders that they might use for different backgrounds (from different docs).
> * text doesn't seem to update live; I'm not sure what's up with that... it's good enough for a first pass but we'll need a ticket to come back.

`ImageCustomizationContext` is a new thing? Do you mean like specifying a color in Figma to use in the shader? We have one called `ImageContext` already.

File: designcompose/src/main/java/com/android/designcompose/CustomizationContext.kt [hunk 4]
```diff
@@ -102,6 +107,7 @@ private fun Customization.clone(): Customization {
     c.text = text
     c.image = image
     c.imageWithContext = imageWithContext
+    c.brush = brush
```
Comment (rylin8) [inline designcompose/src/main/java/com/android/designcompose/CustomizationContext.kt:110]: I think textFunction (which I probably missed) and brushFunction need to be cloned here also.

File: integration-tests/validation/src/main/java/com/android/designcompose/testapp/validation/MainActivity.kt (hunk 3 above)
Comment (iamralpht) [reply integration-tests/validation/src/main/java/com/android/designcompose/testapp/validation/MainActivity.kt:1780]: I can do a cherry-pick and rework it if needed.

**Comment by iamralpht** [comment]
> `ImageCustomizationContext` is a new thing? Do you mean like specifying a color in Figma to use in the shader? We have one called `ImageContext` already.

No, I meant `ImageContext`, doh. I think I should probably use `ImageContext` instead of nothing...

File: designcompose/src/main/java/com/android/designcompose/FrameRender.kt (hunk 2 above)
Comment (iamralpht) [reply designcompose/src/main/java/com/android/designcompose/FrameRender.kt:333]: Yeah, it's nice how they added `RuntimeShader` and `ShaderBrush` -- made this super simple coz we were already using `Brush` everywhere anyway.

File: docs/_docs/modifiers/modifiers.md (hunk 1 above)
Comment (iamralpht) [reply docs/_docs/modifiers/modifiers.md:125]: I'm living in @timothyfroehlich shadow ;) -- I just followed the amazing docs on testing to generate a baseline image for the shader test. Literally the only manual step I had to do was pull the new `dcf` file off the device, and that's already automated in the `feature/rustlayout` branch :).

File: designcompose/src/main/java/com/android/designcompose/CustomizationContext.kt (hunk 4 above)
Comment (iamralpht) [reply designcompose/src/main/java/com/android/designcompose/CustomizationContext.kt:111]: Yep, good catch! Fixed!

File: integration-tests/validation/src/main/java/com/android/designcompose/testapp/validation/MainActivity.kt (hunk 3 above)
Comment (timothyfroehlich) [reply integration-tests/validation/src/main/java/com/android/designcompose/testapp/validation/MainActivity.kt:1780]: Yes, it's only in the layout branch. The reason is that there is and will be a big diff for lots of things in Validation, so I didn't want to try to continuously handle the conflict between the two. That's part of why we agreed to do the majority of development in the rustlayout branch when we talked a few weeks ago.

**Review by rylin8 — APPROVED** [review]

**Comment by iamralpht** [comment]
ok, filed #572 and #573 for follow-up work.

**Pull request merged**